src/main/java/com/redhat/devtools/lsp4ij/operations/foldingRange/LSPFoldingRangeBuilder.java

```diff
@@ -0,0 +1,94 @@
+package com.redhat.devtools.lsp4ij.operations.foldingRange;
+
+import com.intellij.lang.ASTNode;
+import com.intellij.lang.folding.CustomFoldingBuilder;
+import com.intellij.lang.folding.FoldingDescriptor;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.redhat.devtools.lsp4ij.LSPFileSupport;
+import com.redhat.devtools.lsp4ij.LSPIJUtils;
+import org.eclipse.lsp4j.FoldingRange;
+import org.eclipse.lsp4j.Position;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static com.redhat.devtools.lsp4ij.internal.CompletableFutures.isDoneNormally;
+
+public class LSPFoldingRangeBuilder extends CustomFoldingBuilder {
+
+    @Override
+    protected void buildLanguageFoldRegions(@NotNull List<FoldingDescriptor> descriptors,
+                                            @NotNull PsiElement root,
+                                            @NotNull Document document, boolean quick) {
+        // if quick flag is set, we do nothing here
+        if (quick) {
+            return;
+        }
+        PsiFile file = root.getContainingFile();
+        if (file == null) {
+            return;
+        }
+        LSPFoldingRangeSupport foldingRangeSupport = LSPFileSupport.getSupport(file).getFoldingRangeSupport();
+        foldingRangeSupport.cancel();

```

Comment:
why cancel now?

Reply:
@fbricon you are right, cancel should not done since it should be done only when PsiFile has changed (this behavior is managed by AbstractLSPFeatureSupport).

I removed it.